File: backend/app/schemas/memories.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field

EventType = Literal["anniversary", "import_event", "semantic_event"]
MemoryStatus = Literal["pending", "complete", "failed", "empty"]
RunStatus = Literal["running", "complete", "failed"]


class MemoryImageItem(BaseModel):
    """A single curated image within a memory."""

    id: str
    path: str
    thumbnailPath: Optional[str] = None
    captured_at: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    isFavourite: bool = False
    sort_order: int
    score: Optional[float] = None


class MemoryVideoItem(BaseModel):
    """A short clip curated into a memory, sharing the images' sort_order."""

    id: str
    path: str
    thumbnailPath: Optional[str] = None
    captured_at: Optional[str] = None
    duration: Optional[float] = None
    isFavourite: bool = False
    sort_order: int
    score: Optional[float] = None


class MemoryCard(BaseModel):
    """Memory summary, without the image set. Used by the history filmstrip."""

    memory_id: str
    dedupe_key: str
    event_type: EventType
    status: MemoryStatus
    title: str
    subtitle: Optional[str] = None
    place_label: Optional[str] = None
    center_lat: Optional[float] = None
    center_lon: Optional[float] = None
    surface_date: str
    period_start: Optional[str] = None
    period_end: Optional[str] = None
    image_count: int = 0
    video_count: int = 0
    cover_image_id: Optional[str] = None
    cover_thumbnail_path: Optional[str] = None
    score: float = 0.0
    notified_at: Optional[str] = None
    viewed_at: Optional[str] = None
    dismissed: bool = False
    created_at: Optional[str] = None


class MemoryStory(MemoryCard):
    """A memory with its full image set, for the story viewer."""

    images: List[MemoryImageItem] = Field(default_factory=list)
    # Separate from images because they are separate tables; sort_order runs
    # across both, so the viewer merges them into one sequence.
    videos: List[MemoryVideoItem] = Field(default_factory=list)
    # Per-signal breakdown kept for debugging why an image was selected. Not
    # surfaced in the UI.
    signals: Optional[Dict[str, Any]] = None


class GenerateMemoriesRequest(BaseModel):
    """Request model for triggering a curation run."""

    force: bool = False
    reference_date: Optional[str] = None


class GenerateMemoriesData(BaseModel):
    """Outcome of a generate call."""

    run_date: str
    status: RunStatus
    queued: bool


class UpdateMemoryRequest(BaseModel):
    """Request model for updating a memory's user-facing state."""

    viewed: Optional[bool] = None
    dismissed: Optional[bool] = None
    notified: Optional[bool] = None


class MemoryStatusData(BaseModel):
    """Scheduler-facing snapshot. Polled by the Tauri background task."""

    run_date: str
    run_status: Optional[RunStatus] = None
    # Identifies *which* run is being reported. Without it a caller cannot
    # tell its own queued run from one that finished earlier the same day,
    # since run_date is only ever today.
    run_started_at: Optional[str] = None
    indexing_busy: bool = False
    unviewed_count: int = 0
    latest_memory_id: Optional[str] = None
    memories_enabled: bool = True
    notifications_enabled: bool = True


class GetMemoriesData(BaseModel):
    """Paginated memory cards."""

    memories: List[MemoryCard] = Field(default_factory=list)
    total_count: int = 0


class GetMemoryData(BaseModel):
    """A single memory story."""

    memory: MemoryStory


class GetTodayMemoryData(BaseModel):
    """
    The memory to surface now, if any.

    `memory` is null rather than a 404 when nothing qualifies: this endpoint is
    polled, and an empty library is a normal state, not an error.
    """

    memory: Optional[MemoryStory] = None


class DeleteMemoryData(BaseModel):
    """Outcome of a delete call."""

    memory_id: str


class GenerateMemoriesResponse(BaseModel):
    success: bool
    message: str
    data: GenerateMemoriesData


class GetMemoriesResponse(BaseModel):
    success: bool
    message: str
    data: GetMemoriesData


class GetMemoryResponse(BaseModel):
    success: bool
    message: str
    data: GetMemoryData


class GetTodayMemoryResponse(BaseModel):
    success: bool
    message: str
    data: GetTodayMemoryData


class MemoryStatusResponse(BaseModel):
    success: bool
    message: str
    data: MemoryStatusData


class UpdateMemoryResponse(BaseModel):
    success: bool
    message: str
    data: GetMemoryData


class DeleteMemoryResponse(BaseModel):
    success: bool
    message: str
    data: DeleteMemoryData


class ErrorResponse(BaseModel):
    """Error response model"""

    success: bool
    error: str
    message: str

File: frontend/src/api/api-functions/memories.ts
import { memoriesEndpoints } from '../apiEndpoints';
import { apiClient } from '../axiosConfig';
import type { BackendRes } from '@/hooks/useQueryExtension';

// Data Types

export type MemoryEventType = 'anniversary' | 'import_event' | 'semantic_event';

export type MemoryStatus = 'pending' | 'complete' | 'failed' | 'empty';

export type MemoryRunStatus = 'running' | 'complete' | 'failed';

/** A single curated image within a memory. */
export interface MemoryImage {
  id: string;
  path: string;
  thumbnailPath: string | null;
  captured_at: string | null;
  latitude: number | null;
  longitude: number | null;
  isFavourite: boolean;
  sort_order: number;
  score: number | null;
}

/** Memory summary, without the image set. Backs the grid and filmstrip. */
export interface MemoryCard {
  memory_id: string;
  dedupe_key: string;
  event_type: MemoryEventType;
  status: MemoryStatus;
  title: string;
  subtitle: string | null;
  place_label: string | null;
  center_lat: number | null;
  center_lon: number | null;
  surface_date: string;
  period_start: string | null;
  period_end: string | null;
  image_count: number;
  cover_image_id: string | null;
  cover_thumbnail_path: string | null;
  score: number;
  notified_at: string | null;
  viewed_at: string | null;
  dismissed: boolean;
  created_at: string | null;
}

/** A memory with its full image set, for the story viewer. */
export interface MemoryStory extends MemoryCard {
  images: MemoryImage[];
  signals: Record<string, number> | null;
}

export interface MemoryStatusData {
  run_date: string;
  run_status: MemoryRunStatus | null;
  indexing_busy: boolean;
  unviewed_count: number;
  latest_memory_id: string | null;
  memories_enabled: boolean;
  notifications_enabled: boolean;
}

export interface GenerateMemoriesRequest {
  force?: boolean;
  reference_date?: string;
}

export interface ListMemoriesParams {
  limit?: number;
  offset?: number;
  event_type?: MemoryEventType;
  include_viewed?: boolean;
  include_dismissed?: boolean;
}

export interface UpdateMemoryRequest {
  viewed?: boolean;
  dismissed?: boolean;
  notified?: boolean;
}

// API Functions

/** List memory cards, newest first. */
export const getMemories = async (
  params?: ListMemoriesParams,
): Promise<BackendRes<{ memories: MemoryCard[]; total_count: number }>> => {
  const response = await apiClient.get(memoriesEndpoints.list, { params });
  return response.data;
};

/** Get the memory to surface now. `data.memory` is null when there is none. */
export const getTodayMemory = async (): Promise<
  BackendRes<{ memory: MemoryStory | null }>
> => {
  const response = await apiClient.get(memoriesEndpoints.today);
  return response.data;
};

/** Get a single memory with its full image set. */
export const getMemory = async (
  memoryId: string,
): Promise<BackendRes<{ memory: MemoryStory }>> => {
  const response = await apiClient.get(memoriesEndpoints.byId(memoryId));
  return response.data;
};

/** Queue a curation run. */
export const generateMemories = async (
  request: GenerateMemoriesRequest = {},
): Promise<
  BackendRes<{ run_date: string; status: MemoryRunStatus; queued: boolean }>
> => {
  const response = await apiClient.post(memoriesEndpoints.generate, request);
  return response.data;
};

/** Get the scheduler snapshot. */
export const getMemoryStatus = async (): Promise<
  BackendRes<MemoryStatusData>
> => {
  const response = await apiClient.get(memoriesEndpoints.status);
  return response.data;
};

/** Update a memory's viewed, dismissed or notified state. */
export const patchMemory = async ({
  memoryId,
  ...request
}: UpdateMemoryRequest & {
  memoryId: string;
}): Promise<BackendRes<{ memory: MemoryStory }>> => {
  const response = await apiClient.patch(
    memoriesEndpoints.byId(memoryId),
    request,
  );
  return response.data;
};

/** Delete a memory. The underlying photos are untouched. */
export const deleteMemory = async (
  memoryId: string,
): Promise<BackendRes<{ memory_id: string }>> => {
  const response = await apiClient.delete(memoriesEndpoints.byId(memoryId));
  return response.data;
};
